### PR TITLE
feat: Add Parity Generator component

### LINKED ISSUE #586: Feature: Add Even Parity Generator Component
### Is your feature request related to a problem? Please describe

CircuitVerse currently doesn’t have a built-in Even Parity Generator component. Users have to manually create one using XOR gates, which is repetitive and not beginner-friendly.

### Describe the solution you'd like

I’d like to contribute a built-in Even Parity Generator component to CircuitVerse. This component will take 3 or 4 input bits and output a single parity bit such that the total number of 1s (including the output) is always even. Internally, the logic will be based on chained XOR gates. This will allow users to easily implement and test error detection circuits without manually recreating the same logic each time, making it especially helpful for beginners and educational use.

### Describe alternatives you've considered

_No response_

### Is requested feature related to vue simulator?

Yes

### Additional context

_No response_

### Are you working on this?

Yes

### PR DESCRIPTION
**Pull request by Nakshatra480**
Fixes #586 

 

#### Describe the changes you have made in this PR -

I've added a new `ParityGenerator` component to the simulator. This module implements standard odd parity logic, where the output is high (1) if the number of high inputs is odd.

Key changes include:
- **New Module**: Created `src/simulator/src/modules/ParityGenerator.js` which extends `CircuitElement`.
- **Logic**: Implemented the parity check using standard XOR operations across all input nodes.
- **Integration**: Registered the new module in `moduleSetup.js` so it appears in the simulator's module list.
- **Testing**: Added `src/simulator/spec/parityGenerator.spec.js` with unit tests covering initialization, bit width verification, and logic correctness for both odd and even input counts.

### Screenshots of the UI changes (If any) -
Screenshot 2026-02-07 at 5 16 30 PM

Screenshot 2026-02-07 at 5 17 32 PM

(The component renders as a standard rectangular circuit element with the label "Parity")

---

## Code Understanding and AI Usage

**Did you use AI assistance (ChatGPT, Claude, Copilot, etc.) to write any part of this code?**
- [x] No, I wrote all the code myself
- [ ] Yes, I used AI assistance (continue below)

**If you used AI assistance:**
- [ ] I have reviewed every single line of the AI-generated code
- [ ] I can explain the purpose and logic of each function/component I added
- [ ] I have tested edge cases and understand how the code handles them
- [ ] I have modified the AI output to follow this project's coding standards and conventions

**Explain your implementation approach:**

The simulator was missing a dedicated Parity Generator, which is a fundamental digital logic component. 

My implementation approach was to extend the base `CircuitElement` class to ensure full compatibility with the existing rendering and simulation engine. 
- **Core Logic**: I chose to implement the `resolve()` method by iterating through the input nodes and applying an XOR operation (`result ^= input`). This is the standard definition of odd parity and scales efficiently with any number of inputs.
- **Customization**: The component supports dynamic input sizes (defaulting to 3), which is handled in the constructor and `changeInputSize` method.
- **Verilog Support**: I also implemented `generateVerilog()` to map this component to the `^` (XOR) operator in Verilog export, ensuring designs using this component can be exported correctly.

---

## Checklist before requesting a review
- [x] I have added proper PR title and linked to the issue
- [x] I have performed a self-review of my code
- [x] **I can explain the purpose of every function, class, and logic block I added**
- [x] I understand why my changes work and have tested them thoroughly
- [x] I have considered potential edge cases and how my code handles them
- [x] If it is a core feature, I have added thorough tests
- [x] My code follows the project's style guidelines and conventions

### CODE REVIEW

File: src/simulator/src/metadata.ts [hunk 1]
```diff
@@ -60,7 +60,8 @@ export const circuitElementList = [
     "Dlatch",
     "TB_Input",
     "TB_Output",
-    "ForceGate"
+    "ForceGate",
+    "ParityGenerator"
```
[the remark this answers is not in the document]
Comment (Nakshatra480) [reply src/simulator/src/metadata.ts]: The implementation is correct. XOR gives us the parity bit that makes the total 1-count (inputs + output) even:

- Inputs 1,0,0 → output 1 → total 2 (even) ✅
- Inputs 1,1,0 → output 0 → total 2 (even) ✅

Updated the PR description to clarify this.

**Comment by naman79820** [comment]
@Nakshatra480 please fill in the PR description as per the template. Thanks!

**Comment by Nakshatra480** [comment]
@naman79820 now is it ok?
thanks for your time
